File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.java
class Solution {
    public int characterReplacement(String s, int k) {
        /*
        AAABABB
         l    r 
        count
        A:3 maxFreq: 3 maxFreqChar: A
        B: 1 maxFreq: 3 maxFreqChar: A -> k>0: k--
        A: 4 B: 1 maxFreq: 4 
        A: 4 B:2 maxFreq: 4 maxFreqChar: A -> k==0 => res:5
        A: 3 B: 3 maxFreq: 3 maxFeqChar: A        
        */

        int l = 0, r=0;
        int res = 0;
        int maxFreq=0;
        int[] freq = new int[26];
        while(r<s.length()){
            char cur = s.charAt(r);
            freq[cur-'A']++;
            if(freq[cur-'A'] > maxFreq){
                maxFreq = freq[cur-'A'];
            }
            int len = r-l+1;
            if(len-maxFreq <= k){
                res = Math.max(len, res);
            }
            else{
                freq[s.charAt(l++)-'A']--;
            }
            r++;
        }
        return res;

    }
}

File: Data Structures & Algorithms/task-scheduling/submission-2.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # A x x x A x x x A -> (maxFreq - 1) * (n+1) + maxCount
        count = [0] * 26

        for t in tasks:
            count[ord(t) - ord('A')] +=1
        
        maxFreq = max(count)
        maxCount = 0
        for c in count:
            maxCount+=1 if c == maxFreq else 0
        
        return max(len(tasks), (maxFreq - 1) * (n+1) + maxCount)